Sources/SPMClientUtils/ClientUtils.swift

```diff
@@ -0,0 +1,125 @@
+        at path: AbsolutePath,
+        diagnosticObserver: DiagnosticObserver,
+        completion: @escaping(Result<Manifest, Error>) -> Void
+    ) {
+        return loadRootManifest(
+            at: path,
+            observabilityScope: diagnosticObserver.observabilitySystem.topScope,
+            completion: completion
+        )
+    }
+
+    /// Loads a package as the root package of the workspace.
+    ///
+    /// This method is most easily used with `tsc_await` from the `TSCBasic` module in the `swift-tools-support-core` package:
+    ///
+    /// ```swift
+    /// let package = try tsc_await { completion in
+    ///     someWorkspace.loadRootPackage(
+    ///         at: somePath,
+    ///         diagnosticObserver: someObserver,
+    ///         completion: completion
+    ///     )
+    /// }
+    /// ```
+    ///
+    /// - Parameters:
+    ///     - path: The path to the package’s root directory.
+    ///     - diagnosticObserver: An observer to handle diagnostics.
+    ///     - completion: A closure that will be executed when the loading is complete (or has failed).
+    public func loadRootPackage(
+        at path: AbsolutePath,
+        diagnosticObserver: DiagnosticObserver,
+        completion: @escaping(Result<Package, Error>) -> Void
+    ) {
+        return loadRootPackage(
+            at: path,
+            observabilityScope: diagnosticObserver.observabilitySystem.topScope,
+            completion: completion
+        )
+    }
+
+    /// Loads the dependency graph of a package as the root package of the workspace.
+    ///
+    /// - Parameters:
+    ///     - rootPath: The path to the package’s root directory.
+    ///     - explicitProduct: Optional; a string may be passed in order to simulate an explicit product name being supplied to the command line (which may affect the resolution of immediate dependencies by causing the inclusion of an otherwise unused product).
+    ///     - diagnosticObserver: An observer to handle diagnostics.
+    public func loadPackageGraph(
+        rootPath: AbsolutePath,
+        explicitProduct: String? = nil,
+        diagnosticObserver: DiagnosticObserver
+    ) throws -> PackageGraph {
+        return try loadPackageGraph(
+            rootPath: rootPath,
+            explicitProduct: explicitProduct,
+            observabilityScope: diagnosticObserver.observabilitySystem.topScope
+        )
+    }
+}
+
+/// An observer that handles diagnostics emitted while a workspace is loading its components.
+public struct DiagnosticObserver {
+
+    fileprivate let observabilitySystem: ObservabilitySystem
+
+    /// Creates a diagnostic observer.
+    ///
+    /// - Parameters:
+    ///     - handler: A closure that will be executed each time a diagnostic is emitted.
+    ///     - scope: The label of the scope where the diagnostic was emmitted.
+    ///     - diagnostic: The diagnostic message.
+    public init(_ handler: @escaping (_ scope: String, _ diagnostic: String) -> Void) {
```

Comment:
What can a client expect about the concurrency aspects of this closure. Are repeated calls serial or overlapping? Do they occur on the main thread, or a background thread, or anywhere?

(This comment *is* in the right place.)

Reply:
It's a great point that this is something that should be documented.  I don't think there is any general restriction.  For sure they are not necessarily all on the main thread (AFAIK — @tomerd or @neonichu might correct me here), and in general there isn't a guarantee that they're non-overlapping either.  Observers that care about that should use their own serial queue.

Reply:
I went with this for now:

> Diagnostics may be produced from any thread and multiple closure executions may overlap. Be aware of the concurrency ramifications for any context captured by the closure.

Reply:
That sounds like a good way of phrasing it.  Thank you for adding the detailed API documentation, it will be great to make sure that stays current as this interface evolves!  We can later add a .docc with brief higher-level articles.

Reply:
+1